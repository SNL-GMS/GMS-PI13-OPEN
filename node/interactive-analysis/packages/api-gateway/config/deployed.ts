// tslint:disable-next-line: no-default-import
import defaultConfig from './default';

// Overwrites default config for a deployed environment
const deployedConfig = {
  ...defaultConfig,
  // new config name
  configName: 'deployed',
  // session memory is from a DB not in memory
  inMemorySession: false,
  // Sets mock enable to false for all services
  configuration: {
    ...defaultConfig.configuration,
    backend: {
      ...defaultConfig.configuration.backend,
      mock: {
        enable: false
      }
    }
  },
  systemMessage: {
    ...defaultConfig.systemMessage,
    backend: {
      ...defaultConfig.systemMessage.backend,
      mock: {
        enable: false
      }
    }
  },
  referenceStation: {
    ...defaultConfig.referenceStation,
    backend: {
      ...defaultConfig.referenceStation.backend,
      mock: {
        enable: false
      }
    }
  },
  processingStation: {
    ...defaultConfig.processingStation,
    backend: {
      ...defaultConfig.processingStation.backend,
      mock: {
        enable: false
      }
    }
  },
  signalDetection: {
    ...defaultConfig.signalDetection,
    backend: {
      ...defaultConfig.signalDetection.backend,
      mock: {
        enable: false
      }
    }
  },
  qcMask: {
    ...defaultConfig.qcMask,
    backend: {
      ...defaultConfig.qcMask.backend,
      mock: {
        enable: false
      }
    }
  },
  userProfile: {
    ...defaultConfig.userProfile,
    backend: {
      ...defaultConfig.userProfile.backend,
      mock: {
        enable: false
      }
    }
  },
  filterWaveform: {
    ...defaultConfig.filterWaveform,
    backend: {
      ...defaultConfig.filterWaveform.backend,
      mock: {
        enable: false
      }
    }
  },
  event: {
    ...defaultConfig.event,
    backend: {
      ...defaultConfig.event.backend,
      mock: {
        enable: false
      }
    }
  },
  channelSegment: {
    ...defaultConfig.channelSegment,
    backend: {
      ...defaultConfig.channelSegment.backend,
      mock: {
        enable: false
      }
    }
  },
  fk: {
    ...defaultConfig.fk,
    backend: {
      ...defaultConfig.fk.backend,
      mock: {
        enable: false
      }
    }
  },
  dataAcquisition: {
    ...defaultConfig.dataAcquisition,
    backend: {
      ...defaultConfig.dataAcquisition.backend,
      mock: {
        enable: false
      }
    }
  },
  performanceMonitoring: {
    ...defaultConfig.performanceMonitoring,
    backend: {
      ...defaultConfig.performanceMonitoring.backend,
      mock: {
        enable: false
      }
    }
  },
  // Change test data location for mock mode - deprecated?
  testData: {
    ...defaultConfig.testData,
    standardTestDataSet: {
      ...defaultConfig.testData.standardTestDataSet,
      stdsDataHome: '/opt/app-root/src/Test_Data_Sets/Standard_Test_Data/'
    }
  },
  // Changes kafka brokers to be the deployed kafka containers
  kafka: {
    ...defaultConfig.kafka,
    brokers: ['kafka1:9092', 'kafka2:9092', 'kafka3:9092']
  }
};

// tslint:disable-next-line: no-default-export
export default deployedConfig;
