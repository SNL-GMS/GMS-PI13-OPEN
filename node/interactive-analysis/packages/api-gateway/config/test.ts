// tslint:disable-next-line: no-default-import
import defaultConfig from './default';

const testConfig = {
  ...defaultConfig,
  // New config name
  configName: 'test',
  // log level for testing is error only
  logLevel: 'error',
  // simulate waveforms
  waveform: {
    ...defaultConfig.waveform,
    waveformMode: 'simulated'
  },
  // change test data directory
  testData: {
    ...defaultConfig.testData,
    standardTestDataSet: {
      ...defaultConfig.testData.standardTestDataSet,
      stdsDataHome: 'resources/test_data/unit-test-data/Standard_Test_Data'
    }
  }
};

// tslint:disable-next-line: no-default-export
export default testConfig;
