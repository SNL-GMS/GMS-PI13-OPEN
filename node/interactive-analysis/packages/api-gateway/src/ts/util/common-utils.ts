import { getRandomInRange, getSecureRandomNumber } from '@gms/common-util';
import { DataPayload } from '../cache/model';
import { Event } from '../event/model-and-schema/model';
import { QcMask } from '../qc-mask/model';
import { SignalDetection } from '../signal-detection/model';

// Constants used in random lat/lon
const LAT_RANGE = 90;
const LON_RANGE = 180;
const FIXED_DECIMAL = 3;
const OFFSET_MIN = 0.1;
const OFFSET_MAX = 2;
const RADIUS = 2;
const RES_RANGE = 4;

/**
 * Returns a random latitude from -90 to 90
 */
export function getRandomLatitude() {
  return getRandomInRange(-LAT_RANGE, LAT_RANGE, FIXED_DECIMAL);
}

/**
 * Returns a random longitude from -180 to 180
 */
export function getRandomLongitude() {
  return getRandomInRange(-LON_RANGE, LON_RANGE, FIXED_DECIMAL);
}
/**
 * Gets random residual between -4 and 4
 */
export function getRandomResidual() {
  return getRandomInRange(-RES_RANGE, RES_RANGE, FIXED_DECIMAL);
}
/**
 * Returns a small offset used in randomizing event location around a station
 */
export function getRandomOffset() {
  const sign = getSecureRandomNumber() < OFFSET_MAX ? -1 : 1;
  return getRandomInRange(OFFSET_MIN, OFFSET_MAX, FIXED_DECIMAL) * sign;
}

/**
 * Returns a point on a circle RADIUS away
 */
export function getRandomLatLonOffset() {
  const angle = getSecureRandomNumber() * Math.PI * 2;
  const x = Math.cos(angle) * RADIUS;
  const y = Math.sin(angle) * RADIUS;
  return {
    lat: x,
    lon: y
  };
}

/**
 * Checks if the object is empty by checking how many keys are present
 * @param object object to check for empty
 * @returns a boolean
 */
export function isObjectEmpty(object: any): boolean {
  return Object.keys(object).length <= 0;
}

/**
 * Walk thru the double array calling fixNaNValues for each row
 * @param values the values
 */
export function fixNaNValuesDoubleArray(values: number[][]) {
  values.forEach(fixNanValues);
}

/**
 * Walks the array and replaces any NaN values with undefined
 * @param values array of numbers
 */
export function fixNanValues(values: number[]) {
  values.forEach((val, index) => {
    if (val !== undefined && isNaN(val)) {
      values[index] = undefined;
    }
  });
}

/**
 * Creates an empty data payload with the given update type
 * @param updateType the update type
 */
export function createEmptyDataPayload(): DataPayload {
  return {
    events: Object.seal([]),
    sds: Object.seal([]),
    qcMasks: Object.seal([])
  };
}

/**
 * Creates a data payload object with the given data
 * @param events events in payload
 * @param sds sds in payload
 * @param qcMasks qc masks in payload
 * @param updateType optional update type, defaults to full event update
 */
export function createDataPayload(
  events: Event[],
  sds: SignalDetection[],
  qcMasks: QcMask[]
): DataPayload {
  return {
    events: Object.seal(events),
    sds: Object.seal(sds),
    qcMasks: Object.seal(qcMasks)
  };
}

/**
 * Replaces an item in the array if the items id is already in the list
 * @param list array of items
 * @param itemToReplaceWithOrAdd item to updated by id in the list
 */
export function replaceByIdOrAddToList<T extends { id: string }>(
  collection: any,
  itemToReplaceWithOrAdd: T
): T[] {
  const list = collection as T[];
  const indexToReplace = list.findIndex(item => item.id === itemToReplaceWithOrAdd.id);
  if (indexToReplace >= 0) {
    list[indexToReplace] = itemToReplaceWithOrAdd;
  } else {
    list.push(itemToReplaceWithOrAdd);
  }
  return list;
}
