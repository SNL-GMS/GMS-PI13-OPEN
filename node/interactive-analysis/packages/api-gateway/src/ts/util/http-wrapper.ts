import { jsonPretty } from '@gms/common-util';
import { ApolloError } from 'apollo-server-core';
import Axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios';
import clone from 'lodash/clone';
import toLower from 'lodash/toLower';
import msgpack from 'msgpack-lite';
import { gatewayLogger as logger } from '../log/gateway-logger';

/**
 * Http Response
 */
// tslint:disable-next-line: no-empty-interface
export interface HttpResponse<T = any> extends AxiosResponse<T> {}

/**
 * Shortens and formats a string with a ellipsis.
 *
 * @param str the string
 * @param length the desired maximum length of the string
 */
const ellipsis = (str: string, length: number = 500) =>
  length === undefined || str.length > length ? str.substr(0, length) + '...' : str;

/**
 * Returns a formated string from the HTTP Response status text.
 * @param response the http response
 * @param length (optional) the maximum length of the formatted string
 */
export const httpResponseFormattedStatusText = <T>(
  response: HttpResponse<T>,
  length: number = 500
): string => ellipsis(response.statusText, length);

/**
 * Returns a formated string from the HTTP Response.
 * @param response the http response
 */
export const httpResponseFormattedString = <T>(response: HttpResponse<T>): string =>
  `HttpResponse: ` +
  `call=${response.request.url || response.request.method || response.request.baseUrl} ` +
  `status=${response.status} `;

/**
 * Returns true if the response was successful; false otherwise
 * @param response the http response
 */
export const isHttpResponseError = <T>(response: HttpResponse<T>): boolean =>
  // tslint:disable-next-line: no-magic-numbers
  response.status !== 200;

/**
 * A lightweight wrapper around the Axios mock adapter library that provides convenience
 * wrapper functions to register mock handlers around simulated gms backend services.
 */
export class HttpMockWrapper {
  /** Mocks */
  private readonly mocks: Map<string, <I, T>(input: I) => Promise<T>> = new Map();

  /**
   * Initialize the axios mock wrapper with an instance of the axios mock adapter.
   * @param mockAdapter The axios mock adapter to wrap
   */
  public constructor() {
    this.mocks = new Map();
  }

  /**
   * Checks if mock url exists
   * @param url url
   * @returns a boolean
   */
  public has(url: string) {
    return this.mocks.has(url);
  }

  /**
   * Removes the mock handles
   * @param url url
   */
  public remove(url: string) {
    logger.debug(`Removing mock handlers : ${url}`);
    this.mocks.delete(url);
  }

  /**
   * Clears the mock handles
   */
  public clear(): void {
    logger.debug(`Removing all mock handlers.`);
    this.mocks.clear();
  }

  /**
   * Configure a mock service interface at the provided URL, with the provided handler function.
   * @param url The URL to mock a POST service interface for
   * @param handler The handler function to be called whenever a request is sent to the input URL.
   * This function should accept an input object representing the parsed JSON request body, and should
   * return an object representing the result to encode as the HTTP response body.
   */
  public onMock(url: string, handler: (input?: any) => any): void {
    logger.info(`Registering mock handler for url ${url}`);
    this.mocks.set(
      url,
      async (input?: any): Promise<any> =>
        // tslint:disable-next-line: no-inferred-empty-object-type
        new Promise((resolve: any) => {
          resolve(handler(input));
        }).catch(error => {
          logger.error(`Error in mock response url: ${url}  input: ${ellipsis(jsonPretty(input))}`);
          logger.error(`Error in mock response: ${error}`);
          throw new ApolloError(`Service error for url: "${url}" error: ${error}`);
        })
    );
  }

  /**
   * Http request helper method
   * @param url the url
   * @param data request body
   * @returns request response as promise
   */
  public async request<T>(url: string, data?: any): Promise<HttpResponse<T>> {
    if (this.has(url)) {
      const action = this.mocks.get(url);
      const methodReturn: Promise<HttpResponse<T>> = action(data).then((result: T) => ({
        data: result,
        status: 200,
        statusText: 'Mocked Successful',
        headers: {
          'content-type': 'application/json'
        },
        config: {
          url,
          data
        },
        request: {}
      }));
      return methodReturn;
    }
  }
}

/**
 * A lightweight wrapper around the Axios HTTP client providing convenience functions for
 * making HTTP GET and POST requests.
 */
export class HttpClientWrapper {
  /** The axios HTTP client instance */
  private readonly axiosInstance: AxiosInstance;

  /** Axios mock http wrapper */
  private mockAdaptor: HttpMockWrapper;

  /**
   * Initialize the axios wrapper with an instance of the axios http client.
   * @param config (optional) configuration settings used to initialize the axios client
   */
  public constructor(config?: AxiosRequestConfig) {
    // TODO consider passing config parameters to the axios.create() method (e.g. base url)
    // tslint:disable-next-line:no-magic-numbers
    this.axiosInstance = Axios.create({ maxContentLength: 100 * 1024 * 1024 });
  }

  /**
   * Handles real and mock http request
   * @param requestConfig request config
   * @param data request body data
   * @returns request result as promise
   */
  public async request<T>(requestConfig: AxiosRequestConfig, data?: any): Promise<HttpResponse<T>> {
    // Throw an error if the request configuration is undefined
    if (!requestConfig) {
      throw new Error('Cannot send HTTP service request with undefined request configuration');
    }
    // Throw an error if the request configuration does not include the url
    if (!requestConfig.url) {
      throw new Error('Cannot send HTTP service request with undefined URL');
    }

    if (this.mockAdaptor && this.mockAdaptor.has(requestConfig.url)) {
      return this.mockAdaptor.request<T>(requestConfig.url, data);
    }

    // Build the HTTP request object from the provided request config and data
    const request = clone(requestConfig);
    // If request data are provided, provide them as 'params' for GET requests or 'data'
    // for all other methods
    if (data) {
      // check if a post or a get
      if (!request.method || toLower(request.method) === 'get') {
        const paramsSerializer = params =>
          params
            ? Object.keys(params)
                .map(key => `${key}/${params[key]}`)
                .join('/')
            : '';
        // !TODO To be fixed in a CR
        // !This should be fixed once the CR for correcting how params are handled for the request
        // !for stations -> '/name/demo' should be something like `?name=demo` or `?name/demo`
        // !unable to use the params serializer, because axios adds a '?' between the url and params
        // GET / params and custom serializer for get parameters
        // request.params = data;
        // requestConfig.paramsSerializer = paramsSerializer;
        request.url = request.url.concat(paramsSerializer(data));
      } else {
        // POST / data
        // If request content-type header is set to msgpack,
        // encode the request body as msgpack
        const requestType = getHeaderValue(request, 'content-type');
        const encodedData =
          requestType && requestType === 'application/msgpack'
            ? msgpack.encode(data)
            : requestType && requestType === 'text/plain'
            ? `\"${data}\"`
            : data;

        if (requestType && requestType === 'application/msgpack') {
          logger.debug(
            // tslint:disable-next-line:no-magic-numbers
            `Encoding request as msgpack buffer length: ${encodedData.length / (1024 * 1024)}`
          );
        }
        request.data = encodedData;
      }
    }
    return this.axiosInstance(request)
      .then((response: AxiosResponse<any>) => {
        // If the request is configured to accept msgpack responses,
        // decode the response body from message pack; otherwise return the response
        // body as is (e.g. for JSON or plain text encodings)
        const contentType: string = getHeaderValue(response, 'content-type');
        const isMessagePack: boolean = contentType && contentType === 'application/msgpack';
        return {
          ...response,
          headers: {
            ...response.headers,
            'content-type': isMessagePack ? 'application/json' : contentType
          },
          data: isMessagePack ? (msgpack.decode(response.data) as T) : response.data
        };
      })
      .catch(error => {
        let httpResponse;
        if (error.response) {
          // The request was made and the server responded with a status code outside the range of 2xx

          // some services return a more detailed error response message
          // that has been set to the data field, append this to the statusText
          const statusText: string =
            error.response.data !== undefined &&
            error.response.data !== null &&
            (typeof error.response.data === 'string' || error.response.data instanceof String)
              ? `${error.response.statusText}: ${error.response.data.trim()}`
              : error.response.statusText;

          httpResponse = {
            data: undefined,
            status: error.response.status,
            statusText,
            headers: error.response.headers,
            config: error.response.config,
            request: error.response.request
          };
        } else {
          // The request was made but no response was received
          httpResponse = {
            data: undefined,
            status: 400, // Bad Request
            statusText: error.message,
            headers: {},
            config: error.config,
            request: error.request
          };
        }
        logger.error(
          `HTTP Request Error: `,
          `\n\turl: ${httpResponse.config.url}`,
          `\n\tstatus: ${httpResponse.status}`,
          // TODO update after upgrading to the latest version of winston: there is a method called isDebugEnabled
          `\n\tstatusText: ${
            logger.level === 'debug' ? httpResponse.statusText : ellipsis(httpResponse.statusText)
          }`,
          `\n\trequestData: ${
            logger.level === 'debug' ? jsonPretty(data) : ellipsis(jsonPretty(data))
          }`
        );
        const properties: Map<string, any> = new Map<string, any>();
        properties.set('httpResponse', httpResponse);
        // tslint:disable-next-line: max-line-length
        throw new ApolloError(
          `Service error for url: "${httpResponse.config.url}" error: ${ellipsis(
            httpResponse.statusText
          )}`,
          httpResponse.status,
          properties
        );
      });
  }

  /**
   * Create & return a new AxiosMockWrapper for this client
   */
  public createHttpMockWrapper(): HttpMockWrapper {
    return (this.mockAdaptor = new HttpMockWrapper());
  }
}

/**
 * Gets the header value from request
 * @param httpConfig http config
 * @param headerName header name
 * @returns a value from the header
 */
export function getHeaderValue(httpConfig: any, headerName: string): string {
  let value;
  if (httpConfig && httpConfig.headers) {
    Object.keys(httpConfig.headers).forEach(key => {
      if (key.toLowerCase() === headerName.toLowerCase()) {
        value = httpConfig.headers[key];
      }
    });
  }
  return value;
}
