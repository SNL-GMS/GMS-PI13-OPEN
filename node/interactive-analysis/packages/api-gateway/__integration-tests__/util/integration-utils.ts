import { readJsonData } from '@gms/common-util';
import config from 'config';
import path from 'path';
import { resolveTestDataPaths } from '../../src/ts/util/test-data-util';

// tslint:disable: no-magic-numbers

// ASAR station id
export const STATION_IDS = ['1c4a8d11-178a-3c5d-a902-ac599a155621'];

// Channel ids for fk for MK01 and MK02
export const FK_CHANNEL_IDS = [
  '0d99cb69-425a-35ac-894c-9946637c5342',
  '0d8f3dd8-0f21-3558-9420-9198401de454'
];

// Channel segment ID for ASAR FKb
export const ASAR_FKB_CHANNEL_SEGMENT_ID = '9055926c-0f14-3b4e-8299-c79f3bc2f270';

// KDAK channel ids
export const KDAK_CHANNEL_IDS = [
  '991b0b6c-9708-3f63-98c4-b22ce06448f6',
  '842e91ea-4401-3d36-9a63-0fcc6da11c45'
];

// Event Id
export const EVENT_ID = 'a6bd4ad6-6224-309d-9cb2-866461abe679';

// Common Time range for testing
export const TIME_RANGE = {
  startTime: '2010-05-20T22:00:00Z',
  endTime: '2010-05-20T23:00:00Z',
  shortEndTime: '2010-05-20T22:05:00Z'
};

export const DEFAULT_USER = `"defaultUser"`;

/**
 * Get integration input from resources folder
 */
export function getIntegrationInput(integrationName: string) {
  const integrationPath = resolveTestDataPaths().integrationDataHome;
  const integrationInputs = config.get('testData.integrationInputs');
  return readJsonData(integrationPath.concat(path.sep).concat(integrationInputs[integrationName]));
}
