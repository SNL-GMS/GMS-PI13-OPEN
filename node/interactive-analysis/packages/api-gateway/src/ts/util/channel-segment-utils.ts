import { toEpochSeconds, toOSDTime } from '@gms/common-util';
import isEmpty from 'lodash/isEmpty';
import { UserContext } from '../cache/model';
import {
  ChannelSegment,
  isFkSpectraTimeSeries,
  isWaveformChannelSegment,
  isWaveformTimeSeries,
  OSDChannelSegment,
  OSDTimeSeries,
  TimeSeries
} from '../channel-segment/model';
import { FkPowerSpectra } from '../fk/model';
import { gatewayLogger as logger } from '../log/gateway-logger';
import { Waveform } from '../waveform/model';
import { fixNanValues, fixNaNValuesDoubleArray } from './common-utils';
import { convertFktoOSDfromAPI } from './fk-utils';

/**
 * Converts a channel segment in OSD compatible format to api (graphql) format for the frontend.
 * Ideally this method will go away if/when the COI data structures are finalized
 * and we adjust our data structures to match them.
 * @param osdChannelSegment a channel segment in OSD compatible format
 * @param channel segment in api gateway format
 * @returns a time series channel segment
 */
export function convertChannelSegmentFromOSDToAPI(
  userContext: UserContext,
  osdChannelSegment: OSDChannelSegment<OSDTimeSeries>
): ChannelSegment<TimeSeries> {
  if (
    osdChannelSegment &&
    !isEmpty(osdChannelSegment) &&
    osdChannelSegment.timeseries &&
    !isEmpty(osdChannelSegment.timeseries)
  ) {
    try {
      const apiTimeSeries: TimeSeries[] = osdChannelSegment.timeseries.map(
        (timeseries: OSDTimeSeries) => {
          const ts: TimeSeries = {
            ...timeseries,
            startTime: toEpochSeconds(timeseries.startTime)
          };
          if (isFkSpectraTimeSeries(ts, osdChannelSegment.timeseriesType)) {
            ts.reviewed = false;
            ts.spectrums.forEach(spectrum => {
              fixNaNValuesDoubleArray(spectrum.fstat);
              fixNaNValuesDoubleArray(spectrum.power);
            });
          } else if (isWaveformTimeSeries(ts, osdChannelSegment.timeseriesType)) {
            fixNanValues(ts.values);
          }
          return ts;
        }
      );
      return {
        id: osdChannelSegment.id,
        name: osdChannelSegment.name,
        channel: osdChannelSegment.channel,
        type: osdChannelSegment.type,
        startTime: toEpochSeconds(osdChannelSegment.startTime),
        endTime: toEpochSeconds(osdChannelSegment.endTime),
        timeseriesType: osdChannelSegment.timeseriesType,
        requiresSave: false, // coming from OSD means nothing changed
        timeseries: apiTimeSeries
      };
    } catch (error) {
      logger.error(`Invalid OSD Channel segment data; failed to convert: ${error}`);
      return undefined;
    }
  } else {
    return undefined;
  }
}

/**
 * Converts a channel segment in api (graphql) format to an OSD compatible channel segment format.
 * Ideally this method will go away if/when the COI data structures are finalized
 * and we adjust our data structures to match them.
 * @param channelSegment a channel segment in api gateway format
 * @return a OSD compatible channel segment
 */
export function convertChannelSegmentFromAPIToOSD(
  channelSegment: ChannelSegment<TimeSeries>
): OSDChannelSegment<OSDTimeSeries> {
  let osdTimeSeries: OSDTimeSeries[];
  if (isWaveformChannelSegment(channelSegment)) {
    osdTimeSeries = channelSegment.timeseries.map((waveform: Waveform) => ({
      ...waveform,
      startTime: toOSDTime(waveform.startTime)
    }));
  } else {
    osdTimeSeries = convertFktoOSDfromAPI(channelSegment.timeseries as FkPowerSpectra[]);
  }
  const osdChannelSegment: OSDChannelSegment<OSDTimeSeries> = {
    id: channelSegment.id,
    channel: channelSegment.channel,
    name: channelSegment.name,
    timeseriesType: channelSegment.timeseriesType,
    type: channelSegment.type,
    startTime: toOSDTime(channelSegment.startTime),
    endTime: toOSDTime(channelSegment.endTime),
    timeseries: osdTimeSeries
  };
  return osdChannelSegment;
}

/**
 * Truncates the channel segment timeseries to only be within the requested time range
 * if the channel segment timeseries is partially in and partially out of the requested
 * time range.
 * @param channelSegment channel segment to window
 * @param timeRange the time range tto window within
 * @returns a waveform channel segment
 */
export function truncateChannelSegmentTimeseries(
  channelSegment: ChannelSegment<TimeSeries>,
  startTime: number,
  endTime: number
): ChannelSegment<TimeSeries> {
  // Window truncate the channel segments to the time range if there is overlap
  if (
    channelSegment.timeseries &&
    channelSegment.timeseries.length > 0 &&
    (startTime > channelSegment.startTime || endTime < channelSegment.endTime)
  ) {
    const windowedTimeseries = channelSegment.timeseries.map((timeseries: Waveform) => {
      // Calculate the window for the smaller waveform to add
      const startSampleIndex = (startTime - channelSegment.startTime) * timeseries.sampleRate;
      const endSampleIndex = (endTime - startTime) * timeseries.sampleRate + startSampleIndex;

      // Window the values of the times
      const values = timeseries.values.slice(startSampleIndex, endSampleIndex - 1);
      const value = {
        ...timeseries,
        startTime,
        sampleCount: values.length,
        values
      };
      return value;
    });

    // Replace timeseries with windowed timeseries
    return {
      ...channelSegment,
      startTime,
      endTime,
      timeseries: windowedTimeseries
    };
  }
  return channelSegment;
}
