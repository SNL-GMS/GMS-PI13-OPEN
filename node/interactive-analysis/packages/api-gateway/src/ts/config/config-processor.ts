import config from 'config';
import get from 'lodash/get';
import { CacheProcessor } from '../cache/cache-processor';
import { gatewayLogger as logger } from '../log/gateway-logger';
import { HttpClientWrapper } from '../util/http-wrapper';
import * as configMockBackend from './config-mock-backend';

/**
 * API gateway processor for configuration-related data APIs. This class supports:
 * - data fetching & caching from the backend service interfaces
 * - mocking of backend service interfaces based on test configuration
 * - session management
 * - GraphQL query resolution from the user interface client
 */
export class ConfigProcessor {
  /** The singleton instance */
  private static instance: ConfigProcessor;

  /** Local configuration settings */
  private readonly settings: any;

  /** HTTP client wrapper for communicating with backend services */
  private readonly httpWrapper: HttpClientWrapper;

  /** Boolean that checks if processor has been initialized */
  private readonly isInitialized: boolean = false;

  /**
   * Returns the singleton instance of the cache processor.
   * @returns the instance of the cache processor
   */
  public static Instance(): ConfigProcessor {
    if (ConfigProcessor.instance === undefined) {
      ConfigProcessor.instance = new ConfigProcessor();
      ConfigProcessor.instance.initialize();
    }
    return ConfigProcessor.instance;
  }

  /**
   * Constructor - initialize the ConfigProcessor
   */
  private constructor() {
    // Load configuration settings
    this.settings = config.get('config');

    // Initialize an http client wrapper
    this.httpWrapper = new HttpClientWrapper();
  }

  /**
   * Retrieve a configuration by key
   * @param key The key to retrieve the configuration value for
   */
  public getConfigByKey(key: string): any {
    logger.debug(`Local fetching configuration for key: ${key}`);
    return get(CacheProcessor.Instance().getConfiguration(), key);
  }

  /**
   * Fetch configuration data from the backend for the provided key string.
   * If the provided key is undefined, this function throws an error.
   * This function propagates errors from the underlying HTTP call.
   * @param key The key to retrieve the configuration value for
   */
  public fetchConfigByKey(key: string): any {
    logger.debug(`Fetching configuration for key: ${key}`);

    // Handle invalid input
    if (!key) {
      throw new Error('Cannot fetch configuration for an undefined key');
    }

    // Retrieve the settings for the service
    // const serviceConfig = this.settings.backend.services.configByKey.requestConfig;
    const input: configMockBackend.ConfigKeyInput = { key };
    // Call the service and extract the response data
    return configMockBackend.getConfigByKey(input);
  }

  /**
   * Initialize the configuration processor, fetching configuration data from the backend
   * This function sets up a mock backend if configured to do so.
   */
  private initialize() {
    if (this.isInitialized) {
      return;
    }
    logger.info('Initializing the configuration processor');

    // If service mocking is enabled, initialize the mock backend
    if (this.settings.backend.mock.enable) {
      configMockBackend.initialize(this.httpWrapper.createHttpMockWrapper());
    }

    // TODO: Figure out how to merge the config data into the map initialized in the Global cache
    // Cache configuration data needed to support the interactive analysis UI
    CacheProcessor.Instance().setConfiguration(this.fetchConfigByKey(this.settings.rootKey));
  }
}
