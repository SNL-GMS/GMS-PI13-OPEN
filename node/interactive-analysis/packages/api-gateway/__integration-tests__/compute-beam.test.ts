// tslint:disable: no-console
// tslint:disable: no-magic-numbers
import config from 'config';
import { HttpClientWrapper } from '../src/ts/util/http-wrapper';
import { getIntegrationInput } from './util/integration-utils';

// Initialize an http client
// Retrieve the request configuration for the service call
const requestConfig = config.get('channelSegment.backend.services.computeBeam.requestConfig');
const httpWrapper = new HttpClientWrapper();

beforeAll(() => {
  jest.setTimeout(30000);
});

describe('compute beam tests', () => {
  test('compute beam', async () => {
    const fpinput: any = getIntegrationInput('computeBeam');
    const response = await httpWrapper.request(requestConfig, fpinput);
    expect(response.status).toEqual(200);
    expect(response.data).toBeDefined();
    const beam = response.data[0];
    // Expect a beam to come back
    expect(beam.type).toContain('DETECTION_BEAM');
    // Expect only one beam back
    expect(beam.timeseries).toHaveLength(1);
    const timeseries = beam.timeseries[0];
    // Expect input waveforms length to match output beam length
    expect(timeseries.values.length).toEqual(fpinput.waveforms[0].timeseries[0].values.length);
    // Snapshot just a few values(first 100) (truncated) - full snapshot fails
    const first100 = timeseries.values.slice(0, 100).map(value => value.toFixed(5));
    expect(first100).toMatchSnapshot();
  });
});
