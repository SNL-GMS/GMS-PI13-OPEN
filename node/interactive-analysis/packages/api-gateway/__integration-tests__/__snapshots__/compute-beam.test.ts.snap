// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compute beam tests compute beam 1`] = `
Array [
  "0.08286",
  "0.08582",
  "0.20196",
  "0.30447",
  "0.24806",
  "0.27937",
  "0.22503",
  "0.17389",
  "0.04231",
  "-0.19466",
  "-0.08193",
  "0.15049",
  "0.14326",
  "-0.06716",
  "-0.13891",
  "-0.03535",
  "-0.12870",
  "-0.31964",
  "-0.42750",
  "-0.48952",
  "-0.46953",
  "-0.42207",
  "-0.37408",
  "-0.31870",
  "-0.32088",
  "-0.41466",
  "-0.53547",
  "-0.55573",
  "-0.52473",
  "-0.44395",
  "-0.42508",
  "-0.42551",
  "-0.33688",
  "-0.40108",
  "-0.49627",
  "-0.55440",
  "-0.58532",
  "-0.50035",
  "-0.29299",
  "-0.10878",
  "-0.08473",
  "-0.12180",
  "-0.26344",
  "-0.21140",
  "-0.15596",
  "-0.17838",
  "-0.01542",
  "-0.13026",
  "-0.17901",
  "-0.23621",
  "-0.44856",
  "-0.35869",
  "-0.35061",
  "-0.32333",
  "-0.29501",
  "-0.35097",
  "-0.32844",
  "-0.37087",
  "-0.23528",
  "-0.19784",
  "-0.28647",
  "-0.22049",
  "-0.24840",
  "-0.26511",
  "-0.15352",
  "-0.08104",
  "-0.16615",
  "-0.21143",
  "-0.16549",
  "-0.21068",
  "-0.16915",
  "0.02332",
  "0.08404",
  "0.02121",
  "-0.00316",
  "-0.16600",
  "-0.16404",
  "0.08389",
  "0.11019",
  "0.07499",
  "0.07621",
  "-0.06778",
  "-0.21033",
  "-0.30481",
  "-0.42969",
  "-0.42903",
  "-0.31965",
  "-0.26496",
  "-0.26514",
  "-0.28313",
  "-0.24102",
  "-0.19769",
  "-0.18964",
  "-0.10559",
  "0.01023",
  "0.14668",
  "0.23087",
  "0.34233",
  "0.51565",
  "0.50432",
]
`;
